pandas/util/testing.py

```diff
@@ -1415,27 +1438,37 @@ def assert_equal(left, right, **kwargs):
     """
     __tracebackhide__ = True
 
-    if isinstance(left, pd.Index):
+    if isinstance(left, Index):
+        _check_isinstance(left, right, Index)
+        right = Index(right)
         assert_index_equal(left, right, **kwargs)
-    elif isinstance(left, pd.Series):
+    elif isinstance(left, Series):
+        assert isinstance(right, Series)
         assert_series_equal(left, right, **kwargs)
-    elif isinstance(left, pd.DataFrame):
+    elif isinstance(left, DataFrame):
+        assert isinstance(right, DataFrame)
         assert_frame_equal(left, right, **kwargs)
     elif isinstance(left, IntervalArray):
+        assert isinstance(right, IntervalArray)
         assert_interval_array_equal(left, right, **kwargs)
     elif isinstance(left, PeriodArray):
+        assert isinstance(right, PeriodArray)
         assert_period_array_equal(left, right, **kwargs)
     elif isinstance(left, DatetimeArray):
+        assert isinstance(right, DatetimeArray)
         assert_datetime_array_equal(left, right, **kwargs)
     elif isinstance(left, TimedeltaArray):
+        assert isinstance(right, TimedeltaArray)
         assert_timedelta_array_equal(left, right, **kwargs)
     elif isinstance(left, ExtensionArray):
+        assert isinstance(right, ExtensionArray)
         assert_extension_array_equal(left, right, **kwargs)
     elif isinstance(left, np.ndarray):
+        assert isinstance(right, np.ndarray)
         assert_numpy_array_equal(left, right, **kwargs)
     elif isinstance(left, str):
         assert kwargs == {}
-        return left == right
+        assert left == right
```

Comment:
I'm a little wary of this - you did this because it conflicted with `None` as a return value right? Are you sure this isn't actually expected to be a boolean return by any of the tests? *may* be OK just need to be very careful

Reply:
I got a PR merged a while ago that was about removing the returns from assert_*_equal type functions. The point was that we shouldn't expect an assert function to return anything other than None. This one was left out somehow, but is now being caught by mypy since we've stated that the function returns None. 

All the tests are still passing so this is fine. Also if you look at the structure of this function it's basically 
```
if type1:
     assert_equal_for_type1(left, right)
if type2:
    assert_equal_for_type2(left, right)
....
else:
    raise NotImplemented
``` 
and as long as no exception is raised the function will return None. Don't think it make sense to return a boolean for only one of the types.